core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -572,28 +576,50 @@ public RequestLogBuilder logBuilder() {
         return log;
     }
 
-    TimeoutScheduler responseTimeoutScheduler() {
-        return responseTimeoutScheduler;
+    TimeoutScheduler responseCancellationScheduler() {
+        return responseCancellationScheduler;
+    }
+
+    @Override
+    public void cancel(Throwable cause) {
+        cancellationCause = cause;
```

Comment:
I see. I didn't think it through. I think we can pass the cause to `TimeoutScheduler` like you [suggested](https://github.com/line/armeria/pull/3080#discussion_r494057835), but then should also pass the cause to `TimeoutTask.run(cause)`? That makes sure the cause is correct when the task runs.

```java

// ClientContext
void cancel(Throwable cause) {
    scheduler.finishNow(cause);
}

@Nullable
Throwable cancellationCause() {
    return scheduler.cause(); // null when timeoutTask has not run
}

// TimeoutScheduler
void init(..., Throwable cause /* the initial cause */) { 
    eventLoop.schedule(() -> invokeTimeoutTask(cause),...);
}

void finishNow(Throwable cause) {
    // the scheduled task will be cancelled if not yet executed
    eventLoop.execute(() -> invokeTimeoutTask(cause));
}

void invokeTimeoutTask(Throwable cause) {
    ...
    timeoutTask.run(cause);
    this.cause = cause; // now we set the final cause. this cannot be changed because state=FINISHED
}
```

We can cancel handlers if cause not instanceof TimeoutException.

Reply:
> but then should also pass the cause to TimeoutTask.run(cause)? 

That makes sense. I agree to pass the cause to the task.

> ```
> // TimeoutScheduler
> void init(..., Throwable cause /* the initial cause */) { 
>     eventLoop.schedule(() -> invokeTimeoutTask(cause),...);
> }
> ```

TimeoutScheduler is internal class. IMO, isn't `RequestTimeoutException` sufficient for the initial cause? We can add the initial cause when we need it. 😀

Reply:
But client uses `ResponseTimeoutException` and service uses `RequestTimeoutException`. We can check cause == null when it's consumed and resolve to either `ResponseTimeoutException` or `RequestTimeoutException`, but I think that's more code ... 😄

Reply:
> But client uses ResponseTimeoutException and service uses RequestTimeoutException

Ah, I didn't think of that. In that case, taking initial cause would be good. 👍

Reply:
Great, working on it 🚀